File: D-MASH PWA/not_messenger/js/node_admission_v4.js
'use strict';
// Kpwd is password-equivalent, not PAKE. Never store it outside encrypted Node storage.
(function(global){
 const PROFILE='ARGON2ID_64M_T3_P1_V1',scriptUrl=global.document?.currentScript?.src||global.DMASH_NODE_WORKER_URLS?.admission;
 const text=value=>new TextEncoder().encode(value),hex=bytes=>Array.from(bytes,b=>b.toString(16).padStart(2,'0')).join('');
 const field=(value,length)=>{if(typeof value!=='string'||!(new RegExp('^[0-9a-f]{'+length+'}$')).test(value))throw Error('Invalid admission field');};
 function challengeBytes(challenge,issuer,applicant,transcriptHash){
  if(!challenge||Object.keys(challenge).sort().join(',')!=='epoch,expires_at,nonce,profile,salt,type,version'||challenge.type!=='NODE_PASSWORD_CHALLENGE'||challenge.version!==4||challenge.profile!==PROFILE)throw Error('Unsupported Node password profile');
  const salt=Uint8Array.from(atob(challenge.salt),c=>c.charCodeAt(0));
  if(salt.length!==16||btoa(String.fromCharCode(...salt))!==challenge.salt)throw Error('Invalid Node salt');
  field(challenge.epoch,32);field(challenge.nonce,64);field(issuer,64);field(applicant,64);
  if(issuer===applicant||!(transcriptHash instanceof Uint8Array)||transcriptHash.length!==32||!Number.isSafeInteger(challenge.expires_at)||challenge.expires_at<0)throw Error('Invalid admission context');
  return text('D-MASH|NODE-ADMISSION|V4\0'+[issuer,applicant,hex(transcriptHash),PROFILE,challenge.salt,challenge.epoch,challenge.nonce,challenge.expires_at].join('|'));
 }
 async function passwordProof(key,challenge,session,{now=Math.floor(Date.now()/1000)}={}){
  if(session.version!==4||session.closed||session.localRole!=='NODE'||session.peerRole!=='NODE'||!global.DmashNodeIdentity.verify(session.localId)||!global.DmashNodeIdentity.verify(session.peerId))throw Error('Authenticated Node work required');
  if(!Number.isSafeInteger(now)||!(now<challenge.expires_at&&challenge.expires_at<=now+90))throw Error('Node password challenge expired');
  if(!(key instanceof Uint8Array)||key.length!==32)throw Error('Invalid Node password key');
  const bytes=challengeBytes(challenge,session.peerId,session.localId,session.transcriptHash);
  const cryptoKey=await crypto.subtle.importKey('raw',key,{name:'HMAC',hash:'SHA-256'},false,['sign']);
  const proof=hex(new Uint8Array(await crypto.subtle.sign('HMAC',cryptoKey,bytes)));
  if(session.closed)throw Error('Node session closed');
  return {type:'NODE_PASSWORD_PROOF',version:4,proof};
 }
 let job=null;
 function derivePasswordKey(password,salt,{signal}={}){
  if(typeof password!=='string'||!password||text(password).length>1024||!(salt instanceof Uint8Array)||salt.length!==16)return Promise.reject(Error('Invalid Node password parameters'));
  if(signal?.aborted)return Promise.reject(Error('Node password derivation cancelled'));
  if(job)return Promise.reject(Error('Node password worker busy'));
  if(!global.Worker||!scriptUrl)return Promise.reject(Error('Node password Worker required'));
  return new Promise((resolve,reject)=>{
   const worker=new Worker(scriptUrl);job=worker;
   const finish=(error,key)=>{
    if(job!==worker){key?.fill(0);return;}
    clearTimeout(timer);signal?.removeEventListener('abort',cancel);worker.terminate();job=null;
    if(error){key?.fill(0);reject(error);}else resolve(key);
   };
   const cancel=()=>finish(Error('Node password derivation cancelled'));
   const timer=setTimeout(()=>finish(Error('Node password derivation expired')),90000);
   signal?.addEventListener('abort',cancel,{once:true});
   worker.onerror=()=>finish(Error('Node password Worker failed'));
   worker.onmessage=({data})=>{
    if(!(data?.key instanceof Uint8Array)||data.key.length!==32)finish(Error('Node password derivation failed'));
    else finish(null,data.key);
   };
   worker.postMessage({password,salt});
  });
 }
 function context(session){
  if(session.version!==4||session.closed||session.localRole!=='NODE'||session.peerRole!=='NODE'||session.localId===session.peerId||!global.DmashNodeIdentity.verify(session.localId)||!global.DmashNodeIdentity.verify(session.peerId))throw Error('Authenticated Node work required');
 }
 async function passwordCredential(password,options={}){
  const salt=crypto.getRandomValues(new Uint8Array(16));
  return {profile:PROFILE,salt:btoa(String.fromCharCode(...salt)),epoch:hex(crypto.getRandomValues(new Uint8Array(16))),key:await derivePasswordKey(password,salt,options)};
 }
 class PasswordGate{
  constructor(credential,{clock=()=>Date.now()/1000}={}){
   if(credential?.profile!==PROFILE||!(credential.key instanceof Uint8Array)||credential.key.length!==32)throw Error('Invalid Node password credential');
   this.credential={...credential,key:credential.key.slice()};this.clock=clock;
   this.pending=new Map();this.attempted=new Map();this.authorized=new Set();this.failures=new Map();
   this.overloadUntil=0;this.generation=0;this.closed=false;
  }
  challenge(session){
   context(session);if(this.closed)throw Error('Node gate closed');
   const now=Math.floor(this.clock());
   for(const [peer,row] of this.failures)if(row.expires<=now)this.failures.delete(peer);
   const failure=this.failures.get(session.peerId);
   if(now<this.overloadUntil||(failure&&failure.until>now))throw Error('Node admission cooldown');
   if(!failure&&this.failures.size>=4096)throw Error('Node admission failure quota');
   if(this.attempted.has(session)||this.attempted.size>=256)throw Error('Node admission attempt limit');
   const challenge={type:'NODE_PASSWORD_CHALLENGE',version:4,profile:PROFILE,salt:this.credential.salt,epoch:this.credential.epoch,nonce:hex(crypto.getRandomValues(new Uint8Array(32))),expires_at:now+90};
   challengeBytes(challenge,session.localId,session.peerId,session.transcriptHash);
   const token={challenge,generation:this.generation};
   this.pending.set(session,token);this.attempted.set(session,token);return {...challenge};
  }
  async verify(session,response){
   context(session);if(this.closed)throw Error('Node gate closed');
   const token=this.pending.get(session);this.pending.delete(session);
   if(!token)return false;
   const failed=()=>{
    const now=Math.floor(this.clock());
    if(!this.failures.has(session.peerId)&&this.failures.size>=4096){this.overloadUntil=now+60;return false;}
    const count=Math.min(7,(this.failures.get(session.peerId)?.count||0)+1);
    this.failures.set(session.peerId,{count,until:now+Math.min(60,2**(count-1)),expires:now+600});return false;
   };
   if(Math.floor(this.clock())>=token.challenge.expires_at)return failed();
   if(!response||Object.keys(response).sort().join(',')!=='proof,type,version'||response.type!=='NODE_PASSWORD_PROOF'||response.version!==4)return failed();
   try{
    field(response.proof,64);
    const signature=Uint8Array.from(response.proof.match(/../g),b=>parseInt(b,16));
    const key=await crypto.subtle.importKey('raw',this.credential.key,{name:'HMAC',hash:'SHA-256'},false,['verify']);
    const valid=await crypto.subtle.verify('HMAC',key,signature,challengeBytes(token.challenge,session.localId,session.peerId,session.transcriptHash));
    if(this.closed||session.closed||token.generation!==this.generation||this.attempted.get(session)!==token)return false;
    if(Math.floor(this.clock())>=token.challenge.expires_at||!valid)return failed();
    this.failures.delete(session.peerId);this.authorized.add(session);return true;
   }catch(_){return failed();}
  }
  require(session){context(session);if(this.closed||!this.authorized.has(session))throw Error('Node password admission required');}
  forget(session){this.pending.delete(session);this.attempted.delete(session);this.authorized.delete(session);}
  revokeAll(){this.generation++;this.pending.clear();this.attempted.clear();this.authorized.clear();}
  close(){this.revokeAll();this.closed=true;this.credential.key.fill(0);}
 }
 if(typeof WorkerGlobalScope!=='undefined'&&global instanceof WorkerGlobalScope){
  importScripts('vendor/argon2-bundled.min.js');
  let used=false;
  global.onmessage=async({data})=>{
   if(used)return;used=true;
   try{
    if(typeof data?.password!=='string'||!data.password||text(data.password).length>1024||!(data.salt instanceof Uint8Array)||data.salt.length!==16)throw Error();
    const result=await global.argon2.hash({pass:data.password,salt:data.salt,time:3,mem:65536,parallelism:1,hashLen:32,type:global.argon2.ArgonType?.Argon2id??2});
    const key=new Uint8Array(result.hash);result.hash.fill(0);
    global.postMessage({key},[key.buffer]);
   }catch(_){global.postMessage({error:true});}
  };
 }
 const api=Object.freeze({PROFILE,challengeBytes,passwordProof,derivePasswordKey,passwordCredential,PasswordGate});
 global.DmashNodeAdmissionV4=api;
 if(typeof module!=='undefined')module.exports=api;
})(typeof window!=='undefined'?window:globalThis);

File: D-MASH/client/backend/node_admission_v4.py
"""Opt-in Node v4 admission proofs. Kpwd is password-equivalent, NOT a PAKE.

Only encrypted v4 NODE sessions may use this gate. Routing, registration and
mailbox grants must remain inaccessible until their directional gate completes.
The endpoint/runtime integration and credential provisioning are separate.
"""
import base64
import hmac
import secrets
import time
from nacl.pwhash import argon2id

from .node_registration_v4 import _hex
from .crypto import NodeCryptoManager

PROFILE = 'ARGON2ID_64M_T3_P1_V1'
DOMAIN = b'D-MASH|NODE-ADMISSION|V4\0'


def derive_password_key(password, salt):
    if not isinstance(password, str) or not password or len(password.encode()) > 1024:
        raise ValueError('invalid Node password length')
    if not isinstance(salt, bytes) or len(salt) != 16:
        raise ValueError('invalid Node password salt')
    return argon2id.kdf(32, password.encode(), salt, opslimit=3, memlimit=64 * 1024 * 1024)


def password_credential(password):
    salt = secrets.token_bytes(16)
    return {'profile':PROFILE, 'salt':base64.b64encode(salt).decode(),
            'epoch':secrets.token_hex(16), 'key':derive_password_key(password,salt)}


def _context(session):
    if (session.version != 4 or session.closed or session.local_role != 'NODE'
            or session.peer_role != 'NODE' or not isinstance(session.transcript_hash,bytes)
            or len(session.transcript_hash) != 32):
        raise PermissionError('v4 Node session required')
    _hex(session.local_id,64); _hex(session.peer_id,64)
    if session.local_id == session.peer_id:
        raise PermissionError('self peering refused')
    if not all(NodeCryptoManager.verify_node_pow(value) for value in (session.local_id,session.peer_id)):
        raise PermissionError('Node identity work required')


def proof_bytes(challenge, issuer, applicant, transcript_hash):
    if not isinstance(challenge,dict) or set(challenge) != {'type','version','profile','salt','epoch','nonce','expires_at'}:
        raise ValueError('invalid Node admission challenge')
    if challenge['type'] != 'NODE_PASSWORD_CHALLENGE' or type(challenge['version']) is not int or challenge['version'] != 4 or challenge['profile'] != PROFILE:
        raise ValueError('unsupported Node password profile')
    salt=base64.b64decode(challenge['salt'],validate=True)
    if len(salt)!=16 or base64.b64encode(salt).decode()!=challenge['salt']:
        raise ValueError('invalid Node password salt')
    _hex(challenge['epoch'],32); _hex(challenge['nonce'],64)
    _hex(issuer,64); _hex(applicant,64)
    if issuer==applicant or not isinstance(transcript_hash,bytes) or len(transcript_hash)!=32:
        raise ValueError('invalid admission context')
    if type(challenge['expires_at']) is not int or not 0 <= challenge['expires_at'] <= 2**53-1:
        raise ValueError('invalid admission expiry')
    return DOMAIN + '|'.join((issuer,applicant,transcript_hash.hex(),PROFILE,
                              challenge['salt'],challenge['epoch'],challenge['nonce'],str(challenge['expires_at']))).encode()


def password_proof(key, challenge, session, *, now=None):
    _context(session)
    current=int(time.time()) if now is None else now
    if type(current) is not int or not current < challenge.get('expires_at',0) <= current+90:
        raise PermissionError('Node password challenge expired')
    if not isinstance(key,bytes) or len(key)!=32:
        raise ValueError('invalid Node password key')
    message=proof_bytes(challenge,session.peer_id,session.local_id,session.transcript_hash)
    return {'type':'NODE_PASSWORD_PROOF','version':4,'proof':hmac.digest(key,message,'sha256').hex()}


class PasswordGate:
    def __init__(self, credential, *, clock=time.time):
        self.clock=clock
        self.credential=dict(credential)
        if credential.get('profile')!=PROFILE or not isinstance(credential.get('key'),bytes) or len(credential['key'])!=32:
            raise ValueError('invalid Node password credential')
        self.pending={}
        self.authorized=set()
        self.attempted=set()
        self.failures={}
        self.overload_until=0

    def challenge(self, session):
        _context(session)
        now=int(self.clock())
        self.failures={peer:row for peer,row in self.failures.items() if row['expires']>now}
        if now<self.overload_until:
            raise PermissionError('Node admission cooldown')
        failure=self.failures.get(session.peer_id)
        if failure and failure['until']>now:
            raise PermissionError('Node admission cooldown')
        if not failure and len(self.failures)>=4096:
            raise PermissionError('Node admission failure quota exceeded')
        if session in self.attempted:
            raise PermissionError('Node admission already attempted')
        if len(self.attempted)>=256:
            raise PermissionError('Node admission quota exceeded')
        value={'type':'NODE_PASSWORD_CHALLENGE','version':4,'profile':PROFILE,
               'salt':self.credential['salt'],'epoch':self.credential['epoch'],
               'nonce':secrets.token_hex(32),'expires_at':int(self.clock())+90}
        proof_bytes(value,session.local_id,session.peer_id,session.transcript_hash)
        self.attempted.add(session)
        self.pending[session]=value
        return dict(value)

    def verify(self, session, response):
        _context(session)
        # Consume before verification: no repeated guesses against one challenge.
        challenge=self.pending.pop(session,None)
        if challenge is None:return False
        def failed():
            now=int(self.clock())
            if session.peer_id not in self.failures and len(self.failures)>=4096:
                self.overload_until=now+60
                return False
            count=min(7,self.failures.get(session.peer_id,{}).get('count',0)+1)
            self.failures[session.peer_id]={'count':count,'until':now+min(60,2**(count-1)),'expires':now+600}
            return False
        if int(self.clock())>=challenge['expires_at']:return failed()
        if not isinstance(response,dict) or set(response)!={'type','version','proof'} or response['type']!='NODE_PASSWORD_PROOF' or type(response['version']) is not int or response['version']!=4:
            return failed()
        try:
            _hex(response['proof'],64)
            expected=hmac.digest(self.credential['key'],proof_bytes(challenge,session.local_id,session.peer_id,session.transcript_hash),'sha256').hex()
            if not hmac.compare_digest(expected,response['proof']):return failed()
            self.failures.pop(session.peer_id,None)
            self.authorized.add(session)
            return True
        except (ValueError,TypeError):return failed()

    def require(self,session):
        _context(session)
        if session not in self.authorized:raise PermissionError('Node password admission required')

    def forget(self,session):
        self.pending.pop(session,None);self.authorized.discard(session);self.attempted.discard(session)

    def revoke_all(self):
        self.pending.clear();self.authorized.clear();self.attempted.clear()
